File: go/weightedgraph/dense_graph.go
package weightedgraph

import (
	"fmt"
	"errors"
)

// 稠密图 —— 用邻接矩阵

type DenseGraph struct {
	n        int			// 顶点数
	m   		 int			// 边数
	directed bool     // 是否有向
	g        []([]*Edge)   // 邻接矩阵
}

func NewDenseGraph(n int, directed bool) *DenseGraph {
	graph := &DenseGraph{n: n, m: 0, directed: directed,}
	vertors := make([]([]*Edge), n)

	for i := 0; i < n; i++ {
		vertors[i] = make([]*Edge, n)
	}

	graph.g = vertors
	return graph
}

// 返回图中顶点的数量
func (this *DenseGraph) V() int {
	return this.n
}

func (this *DenseGraph) E() int {
	return this.m
}

func (this *DenseGraph) AddEdge(v, w int, weight float32) error {
	if v < 0 || v >= this.n {
		return errors.New("source vertex invalid")
	}

	if w < 0 || w >= this.n {
		return errors.New("dest vertex invalid")
	}

	// 这里其实不用管重复边的问题了，如果有重复的边就直接替换掉，不过要处理一下 m 的问题
	if !this.HasEdge(v, w) {
		this.m++
	}

	this.g[v][w] = NewEdge(v, w, weight)
	if !this.directed {
		this.g[w][v] = NewEdge(v, w, weight)
	}

	return nil
}

func (this *DenseGraph) HasEdge(v, w int) bool {
	if v < 0 || v >= this.n {
		return false
	}

	if w < 0 || w >= this.n {
		return false
	}

	return this.g[v][w] != nil
}

// 返回所有与 v 相邻的点
func (this *DenseGraph) Vertexs(v int) ([]Edge, error) {
	if v < 0 || v > this.n {
		return nil, errors.New("Invalid vertex")
	}

	slice := []Edge{}
	for i := 0; i < this.n; i++ {
		if this.g[v][i] != nil {
			slice = append(slice, *this.g[v][i])
		}
	}
	return slice, nil
}

func (this *DenseGraph) String() string {
	str := "Adjance Martix"

	for i := 0; i < this.n; i++ {
		str += fmt.Sprintf("\n%d Vertexs:", i)
		for j := 0; j < this.n; j++ {
			if this.g[i][j] == nil {
				str += "  NULL"
			} else {
				str += fmt.Sprintf("  %.2f", this.g[i][j].weight)
			}
		}
	}
	return str
}

File: js/graphMST/denseGraph.js
const Edge = require('./edge');

class DenseGraph {
  /**
   * 创建一个以邻接矩阵为底层实现的稠密图
   * @param {int} v 节点数
   * @param {bool} directed 是否是有向图
   */
  constructor(v, directed) {
    this.v = v;
    this.directed = directed;
    this.e = 0;
    this.g = Array(v);
    for (let i = 0; i < v; i++) {
      this.g[i] = Array(v);
      for (let j = 0; j < v; j++) {
        this.g[i][j] = null;
      }
    }
  }

  // 返回图的节点数
  V() {
    return this.v;
  }

  // 返回图的边数
  E() {
    return this.e;
  }

  /**
   * 节点 v 到节点 w 是否有一条边
   * @param {int} v
   * @param {int} w
   * @returns {boolean}
   */
  HasEdge(v, w) {
    if (v < 0 || v >= this.v || w < 0 || w >= this.v) {
      return false;
    }

    return this.g[v][w] !== null;
  }

  /**
   * 向图中添加一条节点 v 到节点 w，权重为 weight 的边
   * @param {int} v 起始节点
   * @param {int} w 终止节点
   * @param {float} weight 边的权重
   */
  AddEdge(v, w, weight) {
    if (v < 0 || v >= this.v || w < 0 || w >= this.w) {
      throw new Error("Invalid vertex");
    }

    if (this.HasEdge(v, w)) {
      return;
    }

    this.e++;

    this.g[v][w] = new Edge(v, w, weight);

    if (!this.directed && v !== w) {
      this.g[w][v] = new Edge(w, v, weight);
    }
  }

  /**
   * 返回节点 v 的所有邻接边
   * @param {int} v
   * @returns {Array<Edge>} 节点的所有相邻边
   */
  Adj(v) {
    if (v < 0 || v >= this.v) {
      throw new Error("Invalid vertex");
    }

    const result = [];
    for(let i = 0; i < this.g[v].length; i++) {
      if (this.g[v][i] !== null) {
        const edge = this.g[v][i];
        result.push(new Edge(edge.V(), edge.W(), edge.Weight()));
      }
    }

    return result;
  }

  toString() {
    let str = "Dense Graph - Adjacency Matrix";
    for (let i = 0; i < this.v; i++) {
      str += `\nVertex: ${i}  `;
      for (let j = 0; j < this.v; j++) {
        if (this.g[i][j] !== null) {
          str += `  ${this.g[i][j].Weight()}`
        } else {
          str += "  NULL"
        }
      }
    }
    return str;
  }
}

module.exports = DenseGraph;
